Remove pickling from motech module

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
This is part of an effort to replace pickling with json serialization for all celery tasks which will allow us to safely upgrade celery. This was a fairly straightforward module, with a few instances just needing to use an id instead of the db object that could then be refetched, an arg that was already in converted to json, and a date that could be parsed from a string representation just as easily within the task.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3944

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
pass repeater id instead of repeater obj to process_repeater

gherceg committed:
remove pickling for import_patients_with_importer task

gherceg committed:
pass id into process_repeat_record and retry_process_repeat_record

gherceg committed:
remove pickling from send_datasets task

gherceg commented on corehq/motech/repeaters/models.py:
```diff
@@ -1157,9 +1157,9 @@ def already_processed():
 
         if fire_synchronously:
-            task(self)
+            task(self._id)
```
I thought about having these tasks accept both an id or object to optimize the synchronous flow, and am open to it. It felt clunky, but if it is worth it I'm happy to make that change.

kaapstorm replied:
I noticed that `retry_process_repeat_record()` and `process_repeat_record()` are identical, and they both call `_process_repeat_record()`.

What if we ...
1. drop `process_repeat_record()`,
2. rename `_process_repeat_record()` to `process_repeat_record()`,
3. and rename `retry_process_repeat_record()` to `process_repeat_record_async()`?

Then `process_repeat_record_async()` takes an ID, and `process_repeat_record()` is called synchronously with a RepeatRecord instance?

(I suspect the answer might be "no", and probably has something to do with the reason behind why  `retry_process_repeat_record()` and `process_repeat_record()` both exist. (Monitoring?))

I'm not certain about the speed considerations. But I am very much looking forward to the day this is on SQL, and we can stop trying to send repeat records whose repeaters are paused, or whose remote endpoints are offline. Then these speed considerations will become irrelevant.

gherceg replied:
Confirming they are separate for monitoring purposes. I added some docs to those methods to make it more clear so thanks for flagging.

gherceg commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -201,14 +204,15 @@ def _process_repeat_record(repeat_record):
-@task(serializer='pickle', queue=settings.CELERY_REPEAT_RECORD_QUEUE)
-def process_repeater(repeater: SQLRepeater):
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(repeater_id):
```
I'll happily continue using typing here, just wasn't sure what the id would be? Is it just an `int`?

kaapstorm replied:
Yeah, it's a normal int Django model primary key. ... And a type hint would be great, to avoid any confusion with `SQLRepeater.repeater_id`, which is the UUID of a Couch Repeater.

gherceg committed:
add comment documenting duplicate method for monitoring purposes

gherceg committed:
add type hint to process_repeater

kaapstorm committed:
repeater.register() already fires repeat_record

Calling `fire()` on the old revision will cause a conflict when we try
to add a new attempt to the old revision.

kaapstorm committed:
Fetch the revision that was updated

kaapstorm committed:
Missed one

gherceg commented:
Thanks for fixing the test issue @kaapstorm! Really appreciate it.

kaapstorm reviewed (approved):
:+1: Thanks for the note and the type hint
